sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java

```diff
@@ -0,0 +1,109 @@
+     *                                and component resources in the stack.
+     * @param outputs                 the stack outputs
+     */
+    private Stack(
+            String projectName,
+            String stackName,
+            List<ResourceTransformation> resourceTransformations,
+            Output<Map<String, Output<?>>> outputs
+    ) {
+        super(
+                RootPulumiStackTypeName,
+                rootPulumiStackName(projectName, stackName),
+                ComponentResourceOptions.builder().resourceTransformations(
+                        resourceTransformations
+                ).build()
+        );
+        this.outputs = requireNonNull(outputs);
+    }
+
+    private static String rootPulumiStackName(String projectName, String stackName) {
+        return String.format("%s-%s", requireNonNull(projectName), requireNonNull(stackName));
+    }
+
+    @InternalUse
+    public Output<Map<String, Output<?>>> getOutputs() {
+        return this.outputs;
+    }
+
+    @InternalUse
+    public static Function<Supplier<Map<String, Output<?>>>, Stack> factory(
+            String projectName,
+            String stackName,
+            List<ResourceTransformation> resourceTransformations
+    ) {
+        // Stack initialization requires specific sequence of events:
+        // - create a Stack instance (note that the super constructor calls readOrRegisterResource)
+        // - set the Stack instance in the global state (user code will try to reach it via Resource constructor)
+        // - register the Stack outputs in the Pulumi engine
+        // - call the user code callback (asynchronously to avoid any additional cyclic dependency problems)
+        // - complete the Stack initialization (by completing the lazyFuture)
+        return (Supplier<Map<String, Output<?>>> callback) -> {
+            // incomplete future that will be completed after the Stack is constructed
+            var lazyFuture = new CompletableFuture<Map<String, Output<?>>>();
+            // incomplete outputs that will be completed when lazyFuture completes
+            var lazyOutputs = Output.of(lazyFuture);
+            // Stack must be the first Resource to be constructed,
+            // before any user code attempts to construct any Resource
+            var stack = new Stack(projectName, stackName, resourceTransformations, lazyOutputs);
+            // register the new Stack with the engine
+            DeploymentInternal.getInstance().setStack(stack);
+            // register the Stack outputs with the engine
+            DeploymentInternal.getInstance().registerResourceOutputs(stack, lazyOutputs);
+            // run the user code callback wrapped in a future (note that future is eager)
+            var callbackFuture = CompletableFuture.supplyAsync(callback);
+            DeploymentInternal.getInstance().getRunner().registerTask("callback", callbackFuture);
```

Comment:
Perhaps share `DeploymentInternal.getInstance()` or even make it a param so we query the global less.

Reply:
done